core/src/main/scala/kafka/admin/AdminUtils.scala

```diff
@@ -61,40 +61,146 @@ object AdminUtils extends Logging {
+   *                 assignment will be used.
+   * @throws AdminOperationException If rack information is supplied but is incomplete, of if
+   *                                 there is not possible to assign each replica to a unique rack
+   *
    */
   def assignReplicasToBrokers(brokerList: Seq[Int],
                               nPartitions: Int,
                               replicationFactor: Int,
                               fixedStartIndex: Int = -1,
-                              startPartitionId: Int = -1)
+                              startPartitionId: Int = -1,
+                              rackInfo: Map[Int, String] = Map())
   : Map[Int, Seq[Int]] = {
+    val subRackInfo = rackInfo.filterKeys(brokerList.contains(_))
+    if (rackInfo.size > 0 && subRackInfo.size != brokerList.size) {
+      throw new AdminOperationException("Incomplete broker-rack mapping supplied for broker list " + brokerList)
+    }
+    val numRacks = subRackInfo.values.toSet.size
+    val useRackAware = numRacks > 0
+    val arrangedBrokerList = if (useRackAware) interlaceBrokersByRack(subRackInfo) else brokerList
     if (nPartitions <= 0)
       throw new AdminOperationException("number of partitions must be larger than 0")
     if (replicationFactor <= 0)
       throw new AdminOperationException("replication factor must be larger than 0")
-    if (replicationFactor > brokerList.size)
+    if (replicationFactor > arrangedBrokerList.size)
       throw new AdminOperationException("replication factor: " + replicationFactor +
-        " larger than available brokers: " + brokerList.size)
+        " larger than available brokers: " + arrangedBrokerList.size)
     val ret = new mutable.HashMap[Int, List[Int]]()
-    val startIndex = if (fixedStartIndex >= 0) fixedStartIndex else rand.nextInt(brokerList.size)
+    val startIndex = if (fixedStartIndex >= 0) fixedStartIndex else rand.nextInt(arrangedBrokerList.size)
     var currentPartitionId = if (startPartitionId >= 0) startPartitionId else 0
 
-    var nextReplicaShift = if (fixedStartIndex >= 0) fixedStartIndex else rand.nextInt(brokerList.size)
+    var nextReplicaShift = if (fixedStartIndex >= 0) fixedStartIndex else rand.nextInt(arrangedBrokerList.size)
     for (i <- 0 until nPartitions) {
-      if (currentPartitionId > 0 && (currentPartitionId % brokerList.size == 0))
+      if (currentPartitionId > 0 && (currentPartitionId % arrangedBrokerList.size == 0))
         nextReplicaShift += 1
-      val firstReplicaIndex = (currentPartitionId + startIndex) % brokerList.size
-      var replicaList = List(brokerList(firstReplicaIndex))
-      for (j <- 0 until replicationFactor - 1)
-        replicaList ::= brokerList(replicaIndex(firstReplicaIndex, nextReplicaShift, j, brokerList.size))
+      val firstReplicaIndex = (currentPartitionId + startIndex) % arrangedBrokerList.size
+      val leader = arrangedBrokerList(firstReplicaIndex)
+      var replicaList = List(leader)
+      if (!useRackAware) {
```

Comment:
It will be difficult to separate them out as they actually share quite a lot of common logic, specifically around choosing the leader of the partition. The code change may seem a lot but actually very little for the default assignment algorithm other than changing the name of `brokerList` to `arrangedBrokerList`. 

I can try separate out the logic of choosing followers into different functions for default vs. rack aware assignment and see how it looks like.